### PR TITLE
cluster-autoscaler: blocked if an instance is detached from MIG

### PR DESCRIPTION
**Pull request by JulienBalestra**
When an instance is detached from a managed instance group in GCP, the cluster-autoscaler return fast [here](https://github.com/kubernetes/autoscaler/blob/5a5845737328f03b4c1353988aded36fa17cc2ea/cluster-autoscaler/core/static_autoscaler.go#L219) and increase the metric `cluster_autoscaler_errors_total{type:cloudprovidererror}`.

I had some trouble to troubleshoot what caused this issue: I had to patch it to add new logs otherwise it was not possible to understand why the cluster autoscaler was stuck.

Terminating the instance from GCP fixed the issue and unstuck the cluster autoscaler.

As we heavily rely on this component, the current cluster autoscaler behaviour could cause us some issues because it prevents the entire cluster from scaling up/down.

The instance detach use case could happen again for troubleshoot/forensic purpose.

This PR has 2 things:
* add more logs around the error management like it is in other places
* continue on error when processing a node: like when the node is nil.

### CODE REVIEW

File: cluster-autoscaler/core/static_autoscaler.go [hunk 1]
```diff
@@ -216,11 +217,13 @@ func (a *StaticAutoscaler) RunOnce(currentTime time.Time) errors.AutoscalerError
 	nodeInfosForGroups, autoscalerError := getNodeInfosForGroups(
 		readyNodes, a.nodeInfoCache, autoscalingContext.CloudProvider, autoscalingContext.ListerRegistry, daemonsets, autoscalingContext.PredicateChecker, a.ignoredTaints)
 	if autoscalerError != nil {
+		klog.Errorf("Failed to get node infos for groups: %v", autoscalerError)
 		return autoscalerError.AddPrefix("failed to build node infos for node groups: ")
```
Comment (JulienBalestra) [inline cluster-autoscaler/core/static_autoscaler.go]: note: here we add a prefix but this error is never displayed by the caller in the main.

File: cluster-autoscaler/core/utils.go [hunk 2]
```diff
@@ -254,7 +254,9 @@ func getNodeInfosForGroups(nodes []*apiv1.Node, nodeInfoCache map[string]*schedu
 	processNode := func(node *apiv1.Node) (bool, string, errors.AutoscalerError) {
 		nodeGroup, err := cloudProvider.NodeGroupForNode(node)
 		if err != nil {
-			return false, "", errors.ToAutoscalerError(errors.CloudProviderError, err)
+			klog.Warningf("Failed to get nodegroup for node %s: %v", node.Name, err)
+			metrics.RegisterError(errors.ToAutoscalerError(errors.CloudProviderError, err))
```
Comment (losipiuk) [inline cluster-autoscaler/core/utils.go]: I think the fix should go to GCP cloud provider. Where we have more control over which situation we should treat as an error and which not. I do not believe we want to treat any error from `NodeGroupForNode` as "not-found".

t seems to me that to implement the changed behavior we only need to make local changes to GceCache.GetMigForInstance, so we do not return an error if we spot the problematic instance. We should also add negative cache for such broken instances so we do not end up with sending to much API call listing MIGs instances.

Comment (JulienBalestra) [reply cluster-autoscaler/core/utils.go]: @losipiuk I've done the changes about the initial issue.
How would you like to do this negative cache ?
Do we return a custom error and remove the instance from the cache at this moment ?

File: cluster-autoscaler/cloudprovider/gce/cache.go [hunk 3]
```diff
@@ -182,7 +182,8 @@ func (gc *GceCache) GetMigForInstance(instanceRef GceRef) (Mig, error) {
 
 			migRef, found := gc.instanceRefToMigRef[instanceRef]
 			if !found {
-				return nil, fmt.Errorf("instance %+v belongs to unknown mig", instanceRef)
+				klog.Warningf("instance %+v belongs to unknown mig", instanceRef)
+				return nil, nil
```
Comment (losipiuk) [inline cluster-autoscaler/cloudprovider/gce/cache.go]: Please also add the `gc.instancesFromUnknownMigs map[GceRef]bool{}` and put the entry the map here.
At the top of the function in `else` branch of loop from line `154` add checking of this new map. If entry is found return `(nil, nil)` from function immediately.

Also add code for cleaning up the new map together with `gc.instanceRefToMigRef`.

Without the new `gc.instancesFromUnknownMigs` map we will be listing instances of mig, which appears to hold the problematic instance, every loop. That will harm the performance.

File: cluster-autoscaler/core/utils.go (hunk 2 above)
Comment (losipiuk) [reply cluster-autoscaler/core/utils.go]: Thanks. I replied what I meant by the change in `cache.go`

**Comment by losipiuk** [comment]
@JulienBalestra are you planning to address the comments on this one. I heard from other parties which would find this fix useful to them :)

File: cluster-autoscaler/cloudprovider/gce/cache.go (hunk 3 above)
Comment (JulienBalestra) [reply cluster-autoscaler/cloudprovider/gce/cache.go]: @losipiuk I've done the requests changes 👍 
I've done a `map[GceRef]struct{}` which is lighter than a bool one as we just store it as index/set. Let me know if everything's fine

File: cluster-autoscaler/cloudprovider/gce/cache.go [hunk 4]
```diff
@@ -195,10 +201,11 @@ func (gc *GceCache) GetMigForInstance(instanceRef GceRef) (Mig, error) {
 	return nil, nil
 }
 
-func (gc *GceCache) removeInstancesForMig(migRef GceRef) {
+func (gc *GceCache) removeInstancesForMigs(migRef GceRef) {
 	for instanceRef, instanceMigRef := range gc.instanceRefToMigRef {
 		if migRef == instanceMigRef {
 			delete(gc.instanceRefToMigRef, instanceRef)
+			delete(gc.instancesFromUnknownMigs, instanceRef)
```
Comment (losipiuk) [inline cluster-autoscaler/cloudprovider/gce/cache.go]: It should not be needed. No harm though.

Comment (losipiuk) [reply cluster-autoscaler/cloudprovider/gce/cache.go]: I think we should clear the Unknown map in RegenerateInstancesCache.
Without it we have (probably harmless leak) as removeInstancesForMigs will not be called for 
Migs which no longer exist

**Comment by losipiuk** [comment]
@JulienBalestra one more thought on this one. If you had some time would you like to send PR which updates the `cluster-autoscaler/cloudprovider/gce/gce_cloud_provider_test.go` to test the added logic?

**Comment by JulienBalestra** [comment]
@losipiuk Yes will do !

**Comment by losipiuk** [comment]
> Łukasz Osipiuk Yes will do !

THanks!

**Comment by losipiuk** [comment]
> > Łukasz Osipiuk Yes will do !
> 
> THanks!

Is there any progress on that one? :)

**Pull request merged**